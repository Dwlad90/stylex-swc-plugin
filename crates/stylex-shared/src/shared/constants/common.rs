pub(crate) static DEFAULT_INJECT_PATH: &str = "@stylexjs/stylex/lib/stylex-inject";
use once_cell::sync::Lazy;
use phf::phf_set;

// Using MDN data as a source of truth to populate the above sets
// by group in alphabetical order:

pub(crate) static VALID_CALLEES: phf::Set<&'static str> = phf_set! {

  "String", "Number", "Math", "Object", "Array"
};

pub(crate) static INVALID_METHODS: phf::Set<&'static str> = phf_set! {
  "random",
  "assign",
  "defineProperties",
  "defineProperty",
  "freeze",
  "seal",
  "splice",
};

pub(crate) static COMPILED_KEY: &str = "$$css";

pub(crate) static SPLIT_TOKEN: &str = "__$$__";

pub(crate) static ROOT_FONT_SIZE: i8 = 16;

pub(crate) static THEME_NAME_KEY: &str = "__themeName__";

pub(crate) static COLOR_FUNCTION_LISTED_NORMALIZED_PROPERTY_VALUES: Lazy<[&str; 9]> =
  Lazy::new(|| {
    [
      "oklch",
      "lch",
      "oklab",
      "hsla",
      "radial-gradient",
      "hwb",
      "lab",
      "clamp",
      "hsl",
    ]
  });

pub(crate) static COLOR_RELATIVE_VALUES_LISTED_NORMALIZED_PROPERTY_VALUES: Lazy<[&str; 7]> =
  Lazy::new(|| [" a ", " b ", " c ", " l ", " h ", " s ", " w "]);
