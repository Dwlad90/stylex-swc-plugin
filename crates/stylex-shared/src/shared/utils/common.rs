use radix_fmt::radix;
use rustc_hash::{FxHashMap, FxHashSet};
use std::{
  any::type_name,
  collections::hash_map::Entry,
  hash::{DefaultHasher, Hash, Hasher},
  ops::Deref,
  path::PathBuf,
};
use swc_core::{
  atoms::Atom,
  common::{FileName, DUMMY_SP},
  ecma::ast::{
    BinaryOp, Decl, Expr, Ident, ImportDecl, ImportSpecifier, KeyValueProp, MemberExpr, Module,
    ModuleDecl, ModuleExportName, ModuleItem, ObjectLit, Pat, Prop, PropName, PropOrSpread, Stmt,
    VarDeclarator,
  },
};

use crate::shared::{
  constants::messages::ILLEGAL_PROP_VALUE,
  enums::{
    data_structures::top_level_expression::{TopLevelExpression, TopLevelExpressionKind},
    misc::VarDeclAction,
  },
  regex::DASHIFY_REGEX,
  structures::{
    base_css_type::BaseCSSType,
    functions::{FunctionConfigType, FunctionMap, FunctionType},
    state_manager::StateManager,
  },
};

use super::ast::{convertors::transform_shorthand_to_key_values, factories::binding_ident_factory};

pub(crate) fn extract_filename_from_path(path: &FileName) -> String {
  match path {
    FileName::Real(path_buf) => path_buf.file_stem().unwrap().to_str().unwrap().to_string(),
    _ => "UnknownFile".to_string(),
  }
}

pub(crate) fn extract_path(path: &FileName) -> &str {
  match path {
    FileName::Real(path_buf) => path_buf.to_str().unwrap(),
    _ => "UnknownFile",
  }
}

pub(crate) fn extract_filename_with_ext_from_path(path: &FileName) -> Option<&str> {
  match path {
    FileName::Real(path_buf) => Some(path_buf.file_name().unwrap().to_str().unwrap()),
    _ => None,
  }
}

pub fn create_hash(value: &str) -> String {
  radix(murmur2::murmur2(value.as_bytes(), 1), 36).to_string()
}

pub(crate) fn wrap_key_in_quotes(key: &str, should_wrap_in_quotes: bool) -> String {
  if should_wrap_in_quotes {
    format!("\"{}\"", key)
  } else {
    key.to_string()
  }
}

pub fn reduce_ident_count<'a>(state: &'a mut StateManager, ident: &'a Ident) {
  if let Entry::Occupied(mut entry) = state.var_decl_count_map.entry(ident.sym.clone()) {
    *entry.get_mut() -= 1;
  }
}

pub fn increase_member_ident(state: &mut StateManager, member_obj: &MemberExpr) {
  if let Some(obj_ident) = member_obj.obj.as_ident() {
    increase_member_ident_count(state, &obj_ident.sym);
  }
}

pub fn reduce_member_expression_count(state: &mut StateManager, member_expression: &MemberExpr) {
  if let Some(obj_ident) = member_expression.obj.as_ident() {
    reduce_member_ident_count(state, &obj_ident.sym);
  }
}

pub fn reduce_member_ident_count(state: &mut StateManager, ident_atom: &Atom) {
  if let Entry::Occupied(mut entry) = state
    .member_object_ident_count_map
    .entry(ident_atom.clone())
  {
    *entry.get_mut() -= 1;
  }
}
pub fn increase_ident_count(state: &mut StateManager, ident: &Ident) {
  increase_ident_count_by_count(state, ident, 1);
}

pub fn increase_member_ident_count(state: &mut StateManager, ident_atom: &Atom) {
  increase_member_ident_count_by_count(state, ident_atom, 1);
}
pub fn increase_ident_count_by_count(state: &mut StateManager, ident: &Ident, count: i16) {
  let ident_id = &ident.sym;

  *state
    .var_decl_count_map
    .entry(ident_id.clone())
    .or_insert(0) += count;
}

pub fn increase_member_ident_count_by_count(
  state: &mut StateManager,
  ident_atom: &Atom,
  count: i16,
) {
  *state
    .member_object_ident_count_map
    .entry(ident_atom.clone())
    .or_insert(0) += count;
}

pub fn get_var_decl_by_ident<'a>(
  ident: &'a Ident,
  traversal_state: &'a mut StateManager,
  functions: &'a FunctionMap,
  action: VarDeclAction,
) -> Option<VarDeclarator> {
  match action {
    VarDeclAction::Increase => increase_ident_count(traversal_state, ident),
    VarDeclAction::Reduce => reduce_ident_count(traversal_state, ident),
    VarDeclAction::None => {}
  };

  if let Some(var_decl) = get_var_decl_from(traversal_state, ident) {
    return Some(var_decl.clone());
  }

  if let Some(func) = functions.identifiers.get(&ident.sym) {
    match func.as_ref() {
      FunctionConfigType::Regular(func) => match &func.fn_ptr {
        FunctionType::Mapper(func) => {
          let result = func();

          let var_decl = VarDeclarator {
            span: DUMMY_SP,
            name: Pat::from(binding_ident_factory(ident.clone())),
            init: Some(Box::new(result)),
            definite: false,
          };

          return Some(var_decl);
        }
        _ => panic!("Function type not supported"),
      },
      FunctionConfigType::Map(_) => unimplemented!(),
    }
  }

  None
}

pub fn get_import_by_ident<'a>(
  ident: &'a Ident,
  state: &'a StateManager,
) -> Option<&'a ImportDecl> {
  get_import_from(state, ident)
}

pub(crate) fn get_var_decl_from<'a>(
  state: &'a StateManager,
  ident: &'a Ident,
) -> Option<&'a VarDeclarator> {
  state.declarations.iter().find(|var_declarator| {
    if let Pat::Ident(binding_ident) = &var_declarator.name {
      return binding_ident.sym == ident.sym;
    }

    false
  })
}

pub(crate) fn get_import_from<'a>(
  state: &'a StateManager,
  ident: &'a Ident,
) -> Option<&'a ImportDecl> {
  state.top_imports.iter().find(|import| {
    import.specifiers.iter().any(|specifier| match specifier {
      ImportSpecifier::Named(named_import) => {
        named_import.local.sym == ident.sym || {
          if let Some(imported) = &named_import.imported {
            match imported {
              ModuleExportName::Ident(export_ident) => export_ident.sym == ident.sym,
              ModuleExportName::Str(strng) => strng.value == ident.sym,
            }
          } else {
            false
          }
        }
      }
      ImportSpecifier::Default(default_import) => default_import.local.sym == ident.sym,
      ImportSpecifier::Namespace(namespace_import) => namespace_import.local.sym == ident.sym,
    })
  })
}

pub(crate) fn _get_var_decl_by_ident_or_member<'a>(
  state: &'a StateManager,
  ident: &'a Ident,
) -> Option<VarDeclarator> {
  state
    .declarations
    .iter()
    .find(|var_declarator| {
      if let Pat::Ident(binding_indent) = &var_declarator.name {
        if binding_indent.sym == ident.sym {
          return true;
        }
      }

      var_declarator
        .init
        .as_ref()
        .and_then(|init| init.as_call())
        .and_then(|call| call.callee.as_expr())
        .and_then(|callee| callee.as_member())
        .and_then(|member| member.prop.as_ident())
        .is_some_and(|member_ident| member_ident.sym == ident.sym)
    })
    .cloned()
}

pub fn get_expr_from_var_decl(var_decl: &VarDeclarator) -> &Expr {
  match &var_decl.init {
    Some(var_decl_init) => var_decl_init,
    None => panic!("Variable declaration is not an expression"),
  }
}

pub fn evaluate_bin_expr(op: BinaryOp, left: f64, right: f64) -> f64 {
  match &op {
    BinaryOp::Add => left + right,
    BinaryOp::Sub => left - right,
    BinaryOp::Mul => left * right,
    BinaryOp::Div => left / right,
    _ => panic!("Operator '{}' is not supported", op),
  }
}

#[allow(dead_code)]
pub(crate) fn type_of<T>(_: T) -> &'static str {
  type_name::<T>()
}

fn prop_name_eq(a: &PropName, b: &PropName) -> bool {
  match (a, b) {
    (PropName::Ident(a), PropName::Ident(b)) => a.sym == b.sym,
    (PropName::Str(a), PropName::Str(b)) => a.value == b.value,
    (PropName::Num(a), PropName::Num(b)) => (a.value - b.value).abs() < f64::EPSILON,

    (PropName::BigInt(a), PropName::BigInt(b)) => a.value == b.value,
    // Add more cases as needed
    _ => false,
  }
}

pub(crate) fn remove_duplicates(props: Vec<PropOrSpread>) -> Vec<PropOrSpread> {
  let mut set = FxHashSet::default();
  let mut result = vec![];

  for prop in props.into_iter().rev() {
    let key = match &prop {
      PropOrSpread::Prop(prop) => match prop.as_ref() {
        Prop::Shorthand(ident) => ident.sym.clone(),
        Prop::KeyValue(key_val) => match &key_val.key {
          PropName::Ident(ident) => ident.sym.clone(),
          PropName::Str(strng) => strng.value.clone(),
          _ => continue,
        },
        _ => continue,
      },
      _ => continue,
    };

    if set.insert(key) {
      result.push(prop);
    }
  }

  result.reverse();

  result
}

pub(crate) fn deep_merge_props(
  old_props: Vec<PropOrSpread>,
  mut new_props: Vec<PropOrSpread>,
) -> Vec<PropOrSpread> {
  for prop in old_props {
    match prop {
      PropOrSpread::Prop(prop) => match *prop {
        Prop::KeyValue(mut kv) => {
          if new_props.iter().any(|p| match p {
            PropOrSpread::Prop(p) => match p.as_ref() {
              Prop::KeyValue(ref existing_kv) => prop_name_eq(&kv.key, &existing_kv.key),
              _ => false,
            },
            _ => false,
          }) {
            if let Expr::Object(ref mut obj) = *kv.value {
              new_props.push(PropOrSpread::Prop(Box::new(Prop::from(KeyValueProp {
                key: kv.key.clone(),
                value: Box::new(Expr::Object(ObjectLit {
                  span: DUMMY_SP,
                  props: obj.props.clone(),
                })),
              }))));
            }
          } else {
            new_props.push(PropOrSpread::Prop(Box::new(Prop::KeyValue(kv))));
          }
        }
        _ => new_props.push(PropOrSpread::Prop(Box::new(*prop))),
      },
      _ => new_props.push(prop),
    }
  }

  remove_duplicates(new_props.into_iter().rev().collect())
}

pub(crate) fn get_hash_map_difference<K, V>(
  orig_map: &FxHashMap<K, V>,
  compare_map: &FxHashMap<K, V>,
) -> FxHashMap<K, V>
where
  K: Eq + Hash + Clone,
  V: PartialEq + Clone,
{
  let mut diff = FxHashMap::default();

  for (key, value) in orig_map {
    if let Some(map2_value) = compare_map.get(key) {
      if value != map2_value {
        diff.insert(key.clone(), value.clone());
      }
    } else {
      diff.insert(key.clone(), value.clone());
    }
  }

  for (key, value) in compare_map {
    if !orig_map.contains_key(key) {
      diff.insert(key.clone(), value.clone());
    }
  }

  diff
}

pub(crate) fn get_hash_map_value_difference(
  orig_map: &FxHashMap<Atom, i16>,
  map2: &FxHashMap<Atom, i16>,
) -> FxHashMap<Atom, i16> {
  let mut diff = FxHashMap::default();

  for (key, value) in orig_map {
    if let Some(map2_value) = map2.get(key) {
      if value != map2_value {
        diff.insert(key.clone(), value - map2_value);
      }
    } else {
      diff.insert(key.clone(), *value);
    }
  }

  diff
}

pub(crate) fn sum_hash_map_values(
  orig_map: &FxHashMap<Atom, i16>,
  compare_map: &FxHashMap<Atom, i16>,
) -> FxHashMap<Atom, i16> {
  let mut sum_map = FxHashMap::default();

  for (key, value) in orig_map {
    sum_map.insert(key.clone(), *value);
  }

  for (key, value) in compare_map {
    sum_map
      .entry(key.clone())
      .and_modify(|e| *e += value)
      .or_insert(*value);
  }

  sum_map
}

pub(crate) fn dashify(s: &str) -> String {
  DASHIFY_REGEX.replace_all(s, "$1-$2").to_lowercase()
}

pub(crate) fn get_css_value(key_value: KeyValueProp) -> (Box<Expr>, Option<BaseCSSType>) {
  let Some(obj) = key_value.value.as_object() else {
    return (key_value.value, None);
  };

  for prop in obj.props.clone().into_iter() {
    match prop {
      PropOrSpread::Spread(_) => unimplemented!("Spread"),
      PropOrSpread::Prop(mut prop) => {
        transform_shorthand_to_key_values(&mut prop);

        match prop.deref() {
          Prop::KeyValue(key_value) => {
            if let Some(ident) = key_value.key.as_ident() {
              if ident.sym == "syntax" {
                let value = obj.props.iter().find(|prop| {
                  match prop {
                    PropOrSpread::Spread(_) => unimplemented!("Spread"),
                    PropOrSpread::Prop(prop) => {
                      let mut prop = prop.clone();
                      transform_shorthand_to_key_values(&mut prop);

                      match prop.as_ref() {
                        Prop::KeyValue(key_value) => {
                          if let Some(ident) = key_value.key.as_ident() {
                            return ident.sym == "value";
                          }
                        }
                        _ => unimplemented!(),
                      }
                    }
                  }

                  false
                });

                if let Some(value) = value {
                  let result_key_value = value
                    .as_prop()
                    .and_then(|prop| prop.as_key_value())
                    .unwrap();

                  return (result_key_value.value.clone(), Some(obj.clone().into()));
                }
              }
            }
          }
          _ => unimplemented!(),
        }
      }
    }
  }

  (key_value.value, None)
}

pub(crate) fn get_key_values_from_object(object: &ObjectLit) -> Vec<KeyValueProp> {
  let mut key_values = vec![];

  for prop in object.props.iter() {
    match prop {
      PropOrSpread::Spread(_) => unimplemented!("Spread"),
      PropOrSpread::Prop(prop) => {
        let mut prop = prop.clone();

        transform_shorthand_to_key_values(&mut prop);

        match prop.as_ref() {
          Prop::KeyValue(key_value) => {
            key_values.push(key_value.clone());
          }
          _ => panic!("{}", ILLEGAL_PROP_VALUE),
        }
      }
    }
  }
  key_values
}

pub(crate) fn fill_top_level_expressions(module: &Module, state: &mut StateManager) {
  module.clone().body.iter().for_each(|item| match &item {
    ModuleItem::ModuleDecl(ModuleDecl::ExportDecl(export_decl)) => {
      if let Decl::Var(decl_var) = &export_decl.decl {
        for decl in &decl_var.decls {
          if let Some(decl_init) = decl.init.as_ref() {
            state.top_level_expressions.push(TopLevelExpression(
              TopLevelExpressionKind::NamedExport,
              *decl_init.clone(),
              Some(decl.name.as_ident().unwrap().sym.clone()),
            ));
            if !state.declarations.contains(&Box::new(&decl)) {
              state.declarations.push(decl.clone());
            }
          }
        }
      }
    }
    ModuleItem::ModuleDecl(ModuleDecl::ExportDefaultExpr(export_decl)) => {
      if let Some(paren) = export_decl.expr.as_paren() {
        state.top_level_expressions.push(TopLevelExpression(
          TopLevelExpressionKind::DefaultExport,
          *paren.expr.clone(),
          None,
        ));
      } else {
        state.top_level_expressions.push(TopLevelExpression(
          TopLevelExpressionKind::DefaultExport,
          *export_decl.expr.clone(),
          None,
        ));
      }
    }
    ModuleItem::Stmt(Stmt::Decl(Decl::Var(var))) => {
      for decl in &var.decls {
        if let Some(decl_init) = decl.init.as_ref() {
          if decl.name.as_ident().is_some() {
            state.top_level_expressions.push(TopLevelExpression(
              TopLevelExpressionKind::Stmt,
              *decl_init.clone(),
              Some(decl.name.as_ident().unwrap().sym.clone()),
            ));

            if !state.declarations.contains(&Box::new(&decl)) {
              state.declarations.push(decl.clone());
            }
          }
        }
      }
    }
    _ => {}
  });
}

pub(crate) fn gen_file_based_identifier(
  file_name: &str,
  export_name: &str,
  key: Option<&str>,
) -> String {
  let key = key.map_or(String::new(), |k| format!(".{}", k));

  format!("{}//{}{}", file_name, export_name, key)
}

pub(crate) fn hash_f64(value: f64) -> u64 {
  let bits = value.to_bits();
  let mut hasher = DefaultHasher::new();
  bits.hash(&mut hasher);
  hasher.finish()
}

pub(crate) fn round_f64(value: f64, decimal_places: u32) -> f64 {
  let multiplier = 10f64.powi(decimal_places as i32);
  (value * multiplier).round() / multiplier
}

pub(crate) fn _resolve_node_package_path(package_name: &str) -> Result<PathBuf, String> {
  match node_resolve::Resolver::default()
    .with_basedir(PathBuf::from("./cwd"))
    .preserve_symlinks(true)
    .with_extensions([".ts", ".tsx", ".js", ".jsx", ".json"])
    .with_main_fields(vec![String::from("main"), String::from("module")])
    .resolve(package_name)
  {
    Ok(path) => Ok(path),
    Err(error) => Err(format!(
      "Error resolving package {}: {:?}",
      package_name, error
    )),
  }
}

pub(crate) fn normalize_expr(expr: &mut Expr) -> &mut Expr {
  match expr {
    Expr::Paren(paren) => normalize_expr(paren.expr.as_mut()),
    _ => expr,
  }
}

pub(crate) fn sort_numbers_factory() -> impl FnMut(&f64, &f64) -> std::cmp::Ordering {
  |a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal)
}

pub(crate) fn char_code_at(s: &str, index: usize) -> Option<u32> {
  s.chars().nth(index).map(|c| c as u32)
}

pub(crate) fn stable_hash<T: Hash>(t: &T) -> u64 {
  let mut hasher = DefaultHasher::new();
  t.hash(&mut hasher);
  hasher.finish()
}

pub(crate) fn find_and_swap_remove<T, F>(vec: &mut Vec<T>, predicate: F) -> Option<T>
where
  F: Fn(&T) -> bool,
{
  vec
    .iter()
    .position(predicate)
    .map(|index| vec.swap_remove(index))
}
