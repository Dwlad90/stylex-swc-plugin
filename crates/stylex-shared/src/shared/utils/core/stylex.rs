use crate::{
  shared::{
    enums::data_structures::fn_result::FnResult, utils::ast::convertors::string_to_expression,
  },
  transform::styleq::common::styleq,
};

use super::parse_nullable_style::ResolvedArg;

pub(crate) fn stylex(values: &Vec<ResolvedArg>) -> Option<FnResult> {
  let result = styleq(values);

  Some(FnResult::Stylex(string_to_expression(
    result.class_name.as_str(),
  )))
}
