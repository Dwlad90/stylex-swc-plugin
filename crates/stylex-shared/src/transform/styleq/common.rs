use std::collections::BTreeMap;

use indexmap::IndexMap;
use log::warn;
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

use crate::shared::{
  constants::common::COMPILED_KEY,
  enums::data_structures::flat_compiled_styles_value::FlatCompiledStylesValue,
};
use crate::shared::{
  structures::types::FlatCompiledStyles,
  utils::core::parse_nullable_style::{ResolvedArg, StyleObject},
};

pub(crate) struct StyleQResult {
  pub(crate) class_name: String,
  pub(crate) inline_style: Option<FlatCompiledStyles>,
  pub(crate) data_style_src: Option<String>,
}

fn get_hash<T>(obj: T) -> u64
where
  T: Hash,
{
  let mut hasher = DefaultHasher::new();
  obj.hash(&mut hasher);
  hasher.finish()
}

pub(crate) fn styleq(arguments: &[ResolvedArg]) -> StyleQResult {
  let mut class_name = String::default();
  let mut debug_string = String::default();

  if arguments.is_empty() {
    // Early return if there are no arguments
    return StyleQResult {
      class_name,
      inline_style: None,
      data_style_src: None,
    };
  }

  let mut defined_properties = vec![]; // The className and inline style to build up

  let inline_style = None;
  let mut next_cache: Option<IndexMap<u64, (String, Vec<String>, String)>> = Some(IndexMap::new());
  let mut styles = arguments.iter().collect::<Vec<_>>();

  while let Some(possible_style) = styles.pop() {
    let possible_style = match possible_style {
      ResolvedArg::StyleObject(_, _, _) => possible_style,
      ResolvedArg::ConditionalStyle(_, value, _, _, _) => {
        if value.is_some() {
          possible_style
        } else {
          continue;
        }
      }
    };

    match possible_style {
      ResolvedArg::StyleObject(style, _, _) => match style {
        StyleObject::Style(style) => {
          let Some(a) = style.get(COMPILED_KEY) else {
            panic!("Style object does not contain a compiled key")
          };

          if let FlatCompiledStylesValue::Bool(_) | FlatCompiledStylesValue::String(_) = a.as_ref()
          {
            let btree_map: BTreeMap<_, _> = style.iter().collect();

            let style_hash = get_hash(btree_map);
            let mut class_name_chunk = String::default();

            // Build up the class names defined by this object
            if let Some(next_cache) = next_cache.as_mut() {
              if let Some((cached_class_name, cached_properties, cached_debug_string)) =
                next_cache.get(&style_hash)
              {
                class_name_chunk = cached_class_name.clone();
                defined_properties.extend(cached_properties.iter().cloned());
                debug_string = cached_debug_string.clone();
              } else {
                // The properties defined by this object
                let mut defined_properties_chunk: Vec<String> = vec![];

                for (prop, value) in style.iter() {
                  if prop.eq(COMPILED_KEY) {
                    let compiled_key_value = &style[prop];

                    let mut compiled_key_value_is_true = false;

                    if let FlatCompiledStylesValue::Bool(value) = compiled_key_value.as_ref()
                      && *value
                    {
                      compiled_key_value_is_true = true;
                    }

                    if !compiled_key_value_is_true {
                      let compiled_key_string_value = match compiled_key_value.as_ref() {
                        FlatCompiledStylesValue::String(strng) => strng.clone(),
                        other => {
                          let other_debug_info = format!("{:?}", other);
                          let variant_name =
                            other_debug_info.split("::").last().unwrap_or("unknown");

                          unimplemented!(
                            "String conversion not implemented for FlatCompiledStylesValue::{}",
                            variant_name
                          )
                        }
                      };

                      debug_string = if !debug_string.is_empty() {
                        format!("{}; {}", compiled_key_string_value, debug_string)
                      } else {
                        compiled_key_string_value
                      };
                    }

                    continue;
                  }

                  if let FlatCompiledStylesValue::Bool(_) = value.as_ref() {
                    warn!(
                      "styleq: {} typeof {:?} is not \"string\" or \"null\".",
                      prop, "Bool"
                    )
                  }

                  if !defined_properties.contains(prop) {
                    defined_properties.push(prop.clone());
                    defined_properties_chunk.push(prop.clone());

                    if let FlatCompiledStylesValue::String(value) = (**value).clone() {
                      class_name_chunk = if class_name_chunk.is_empty() {
                        value.to_string()
                      } else {
                        format!("{} {}", class_name_chunk, value)
                      };
                    }
                  }
                }

                next_cache.insert(
                  style_hash,
                  (
                    class_name_chunk.clone(),
                    defined_properties_chunk,
                    debug_string.clone(),
                  ),
                );
              }
            }

            if !class_name_chunk.is_empty() {
              class_name = if class_name.is_empty() {
                class_name_chunk.clone()
              } else if !class_name.contains(class_name_chunk.as_str()) {
                format!("{} {}", class_name_chunk, class_name)
              } else {
                class_name
              };
            }
          } else {
            unimplemented!("DYNAMIC: Process inline style object")
          }
        }
        StyleObject::Nullable => panic!("Nullable style object is not allowed in styleq"),
        StyleObject::Other => panic!("Other style object is not allowed in styleq"),
        StyleObject::Unreachable => unreachable!("StyleObject::Unreachable in styleq"),
      },
      _ => unreachable!(),
    };
  }

  StyleQResult {
    class_name,
    inline_style,
    data_style_src: Some(debug_string),
  }
}
