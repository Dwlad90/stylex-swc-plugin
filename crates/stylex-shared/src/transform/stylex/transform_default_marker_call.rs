use swc_core::ecma::ast::CallExpr;
use swc_core::{common::comments::Comments, ecma::ast::Expr};

use crate::shared::{
  transformers::stylex_default_maker::stylex_default_marker,
  utils::validators::{is_default_marker_call, validate_stylex_default_marker_indent},
};
use crate::{StyleXTransform, shared::utils::core::js_to_expr::convert_object_to_ast};

impl<C> StyleXTransform<C>
where
  C: Comments,
{
  pub(crate) fn transform_default_marker_call(&mut self, call: &CallExpr) -> Option<Expr> {
    let is_default_marker_call = is_default_marker_call(call, &self.state);

    if is_default_marker_call {
      validate_stylex_default_marker_indent(call, &mut self.state);

      let marker_result = stylex_default_marker(&self.state.options);

      let marker_ast = convert_object_to_ast(&marker_result);

      Some(marker_ast)
    } else {
      None
    }
  }
}
