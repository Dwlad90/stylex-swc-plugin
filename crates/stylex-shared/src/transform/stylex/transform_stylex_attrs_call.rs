use swc_core::{
  common::comments::Comments,
  ecma::ast::{CallExpr, Expr},
};

use crate::{
  StyleXTransform,
  shared::utils::{
    core::{attrs::attrs, stylex_merge::stylex_merge},
    validators::is_attrs_call,
  },
};

impl<C> StyleXTransform<C>
where
  C: Comments,
{
  pub(crate) fn transform_stylex_attrs_call(&mut self, call: &mut CallExpr) -> Option<Expr> {
    let is_attrs_call = is_attrs_call(call, &self.state);

    if is_attrs_call {
      return stylex_merge(call, attrs, &mut self.state);
    }

    None
  }
}
