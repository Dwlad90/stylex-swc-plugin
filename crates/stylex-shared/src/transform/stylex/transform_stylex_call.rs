use swc_core::{
  common::comments::Comments,
  ecma::ast::{CallExpr, Callee, Expr},
};

use crate::{
  StyleXTransform,
  shared::{
    structures::named_import_source::ImportSources,
    utils::core::{stylex::stylex, stylex_merge::stylex_merge},
  },
};

impl<C> StyleXTransform<C>
where
  C: Comments,
{
  pub(crate) fn transform_stylex_call(&mut self, call: &mut CallExpr) -> Option<Expr> {
    match &call.callee {
      Callee::Expr(expr) => match expr.as_ref() {
        Expr::Ident(ident) => {
          if self
            .state
            .stylex_import
            .contains(&ImportSources::Regular(ident.sym.to_string()))
          {
            if let Some(value) = stylex_merge(call, stylex, &mut self.state) {
              return Some(value);
            }
          }
          None
        }
        _ => None,
      },
      _ => None,
    }
  }
}
