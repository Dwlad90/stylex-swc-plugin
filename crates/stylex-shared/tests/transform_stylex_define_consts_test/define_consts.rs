use stylex_shared::{
  StyleXTransform,
  shared::structures::{
    plugin_pass::PluginPass,
    stylex_options::{ModuleResolution, StyleXOptionsParams},
  },
};
use swc_core::ecma::{
  parser::{Syntax, TsSyntax},
  transforms::testing::test,
};
use swc_core::common::FileName;
use std::path::PathBuf;

test!(
  Syntax::Typescript(TsSyntax {
    tsx: true,
    ..Default::default()
  }),
  |tr| StyleXTransform::new_test_with_pass(
    tr.comments.clone(),
    PluginPass {
      cwd: Some(PathBuf::from("/stylex/packages/")),
      filename: FileName::Real("/stylex/packages/TestTheme.stylex.js".into()),
    },
    Some(&mut StyleXOptionsParams {
      unstable_module_resolution: Some(ModuleResolution {
        r#type: "commonJS".to_string(),
        root_dir: Some("/stylex/packages/".to_string()),
        theme_file_extension: None,
      }),
      ..Default::default()
    })
  ),
  constants_are_unique,
  r#"
        import stylex from 'stylex';
        export const breakpoints = stylex.defineConsts({ padding: '10px' });
      "#
);

test!(
  Syntax::Typescript(TsSyntax {
    tsx: true,
    ..Default::default()
  }),
  |tr| StyleXTransform::new_test_with_pass(
    tr.comments.clone(),
    PluginPass {
      cwd: Some(PathBuf::from("/stylex/packages/")),
      filename: FileName::Real("/stylex/packages/TestTheme.stylex.js".into()),
    },
    Some(&mut StyleXOptionsParams {
      unstable_module_resolution: Some(ModuleResolution {
        r#type: "commonJS".to_string(),
        root_dir: Some("/stylex/packages/".to_string()),
        theme_file_extension: None,
      }),
      ..Default::default()
    })
  ),
  constants_object,
  r#"
        import * as stylex from '@stylexjs/stylex';
        export const breakpoints = stylex.defineConsts({
          sm: '(min-width: 768px)',
          md: '(min-width: 1024px)',
          lg: '(min-width: 1280px)',
        });
      "#
);

test!(
  Syntax::Typescript(TsSyntax {
    tsx: true,
    ..Default::default()
  }),
  |tr| StyleXTransform::new_test_with_pass(
    tr.comments.clone(),
    PluginPass {
      cwd: Some(PathBuf::from("/stylex/packages/")),
      filename: FileName::Real("/stylex/packages/TestTheme.stylex.js".into()),
    },
    Some(&mut StyleXOptionsParams {
      unstable_module_resolution: Some(ModuleResolution {
        r#type: "haste".to_string(),
        root_dir: None,
        theme_file_extension: None,
      }),
      ..Default::default()
    })
  ),
  constants_object_haste,
  r#"
        import * as stylex from '@stylexjs/stylex';
        export const breakpoints = stylex.defineConsts({
          sm: '(min-width: 768px)',
          md: '(min-width: 1024px)',
          lg: '(min-width: 1280px)',
        });
      "#
);

test!(
  Syntax::Typescript(TsSyntax {
    tsx: true,
    ..Default::default()
  }),
  |tr| StyleXTransform::new_test_with_pass(
    tr.comments.clone(),
    PluginPass {
      cwd: Some(PathBuf::from("/stylex/packages/")),
      filename: FileName::Real("/stylex/packages/TestTheme.stylex.js".into()),
    },
    Some(&mut StyleXOptionsParams {
      unstable_module_resolution: Some(ModuleResolution {
        r#type: "commonJS".to_string(),
        root_dir: Some("/stylex/packages/".to_string()),
        theme_file_extension: None,
      }),
      ..Default::default()
    })
  ),
  constant_names_special_characters,
  r#"
        import * as stylex from '@stylexjs/stylex';
        export const sizes = stylex.defineConsts({
          'font-size*large': '18px',
        });
      "#
);

test!(
  Syntax::Typescript(TsSyntax {
    tsx: true,
    ..Default::default()
  }),
  |tr| StyleXTransform::new_test_with_pass(
    tr.comments.clone(),
    PluginPass {
      cwd: Some(PathBuf::from("/stylex/packages/")),
      filename: FileName::Real("/stylex/packages/TestTheme.stylex.js".into()),
    },
    Some(&mut StyleXOptionsParams {
      unstable_module_resolution: Some(ModuleResolution {
        r#type: "commonJS".to_string(),
        root_dir: Some("/stylex/packages/".to_string()),
        theme_file_extension: None,
      }),
      ..Default::default()
    })
  ),
  constant_names_number,
  r#"
        import * as stylex from '@stylexjs/stylex';
        export const levels = stylex.defineConsts({
          1: 'one'
        });
      "#
);