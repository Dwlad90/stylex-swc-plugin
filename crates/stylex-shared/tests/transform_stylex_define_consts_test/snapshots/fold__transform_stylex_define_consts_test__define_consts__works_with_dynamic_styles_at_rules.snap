---
source: crates/stylex-shared/tests/transform_stylex_define_consts_test/define_consts.rs
expression: output
---
import * as stylex from '@stylexjs/stylex';
import { breakpoints } from './constants.stylex';
export const styles = {
    node: (color)=>[
            {
                kMwMTN: "x1pitq50 " + (color != null ? "x3d248p" : color),
                $$css: true
            },
            {
                "--x-4xs81a": color != null ? color : undefined
            }
        ]
};
